export: Translate K()-events to  attacker()

rkunnema opened the pull request:
This PR fixes a bug whereby `tamarin-prover-export` only translated `KU()` to ProVerif's `attacker()`, but not `K()`. It also adds a helper function `isKLogFact` to quickly check if a fact is of the `K()` kind.

rkunnema commented:
For reference there is a small diff in the regressions that could not possibly be caused by the change in the export this PR performs.
```
 Cases: 55  Tried: 55  Errors: 0  Failures: 0
running 'make -j 1 fast-case-studies FAST=y 2>/dev/null' ...
The time changed from        25.61s to      0.03s in /fast-tests/ake/bilinear/Scott_EphkRev_analyzed.spthy
  The steps changed from   15 steps to    8 steps in key_secrecy
The time changed from        18.94s to      0.03s in /fast-tests/ake/bilinear/Scott_analyzed.spthy
  The steps changed from  518 steps to  520 steps in key_secrecy
The time changed from         0.92s to      0.03s in /fast-tests/csf12/UM_PFS_analyzed.spthy
  The steps changed from   10 steps to    8 steps in wPFS_initiator_key
  The steps changed from   10 steps to    8 steps in wPFS_responder_key
  ```
  The last update to `case-studies-regression` is from May so ... should I just push a new version with the timings on my computer?

rsasse commented:
Those results look somewhat ok as they have the same result, however it confuses me that all 3 case studies now take the same, almost-0, time, while before it was 25, 18, and 1 second respectively (so very diverse). Is there possibly some issue with the time measurement on your machine?

The number of steps did not change similarly much, 518->520, 10->8, and 15->8, and those are fine to me.

Also, I understand this is not caused by your change, so thank you very much for bringing this up! I wonder when and how this has happened, particularly the speedup.

rkunnema commented:
Yeah, I am also a bit confused by this. The previous code snippet was from the failed integration test, so computed on github (? I guess) 

I get a slightly different result on my M1, on which I have to use `lts-19.32` because otherwise the dependencies break 
```
❯ python3 regressionTests.py
running 'stack install' ...
running 'make -j 1 fast-case-studies FAST=y 2>/dev/null' ...

--------------------------------------------------------------------------------

The total amount of time  changed from 880s to 3s - this is 0%
The total amount of steps did not change

Time elapsed: 0:06:40s
```

I am not sure what we to make of that.

rkunnema commented:
Oh, and to answer the obvious (and fair) question, yes I deletes `case-studies` beforehand. ;) 

Looking into the first file:
```
analyzed: examples/ake/bilinear/Scott_EphkRev.spthy

  output:          examples/ake/bilinear/Scott_EphkRev.spthy.tmp
  processing time: 0.033268s
  
  key_agreement_reachable (exists-trace): verified (12 steps)
  key_secrecy (all-traces): falsified - found trace (15 steps)
```
I see the steps and the timing of the computation of my M1 matches what github computes. One file just mysteriously having slower verification time – this could occur when someone runs the benchmarks in the background while a paper and compiling a PDF, right? I've done that. :D

Maybe our CI magicians can come up with a way to let the github runner compute the regressions? E.g., put the `case-studies` directory someplace that we can download it to overwrite `case-studies-regression` with?

jdreier commented:
The changes are probably related to the DH-neutral issue, but there also seems to be a problem with the timings (but I don't know where this comes from).

The regression targets are hopelessly outdated because the current script is too permissive and hence people did not update them although changes occurred (see Issue https://github.com/tamarin-prover/tamarin-prover/issues/496). We definitely need to update them, and to update the script to fix the timing and comparison issues.

I think there is a way to obtain the CI output, but I did not have time to look into that.

rkunnema commented:
I am confused. For this commit, I've deleted `case-studies-regression` and moved my freshly computed `case-studies` directory in its place. Locally, I get:
```
❯ python3 regressionTests.py
running 'stack install' ...
running 'make -j 1 fast-case-studies FAST=y 2>/dev/null' ...

--------------------------------------------------------------------------------

The total amount of time  changed from 3s to 2s - this is 82%
The total amount of steps did not change

Time elapsed: 0:05:08s
```
CI gives me:
```
The time changed from         0.03s to      0.05s in /fast-tests/ake/bilinear/Scott_EphkRev_analyzed.spthy
  The steps changed from   15 steps to    8 steps in key_secrecy
The time changed from         0.03s to      0.04s in /fast-tests/ake/bilinear/Scott_analyzed.spthy
  The steps changed from  518 steps to  520 steps in key_secrecy
The time changed from         0.05s to      0.04s in /fast-tests/csf12/UM_PFS_analyzed.spthy
  The steps changed from   10 steps to    8 steps in wPFS_initiator_key
  The steps changed from   10 steps to    8 steps in wPFS_responder_key

--------------------------------------------------------------------------------

The total amount of time  changed from 3s to 3s - this is 111%
The total amount of steps changed from 18858 steps to 18849 steps - this is 99%

Time elapsed: 0:11:31s
```
Both versions run the same Maude. I doubt that the number of steps be architecture dependent.

rkunnema commented:
Hi! Ah, @jdreier I missed your comment. The timing issue could be the one fixed here. 

https://github.com/tamarin-prover/tamarin-prover/pull/517

I'll merge this commit now, as the diff in steps seems to be eplained(-ish ... we should keep an eye on this) now - at least it is not due to this export-related PR. The wishlist for the CI is here https://github.com/tamarin-prover/tamarin-prover/issues/496
